src/hooks/useWatchHistory.ts

```diff
@@ -0,0 +1,123 @@
+import { MutableRefObject, useCallback, useEffect, useLayoutEffect, useMemo, useRef } from 'react';
+
+import type { JWPlayer } from '#types/jwplayer';
+import type { PlaylistItem } from '#types/playlist';
+import useEventCallback from '#src/hooks/useEventCallback';
+import { saveItem } from '#src/stores/WatchHistoryController';
+import { useConfigStore } from '#src/stores/ConfigStore';
+import { useWatchHistoryStore } from '#src/stores/WatchHistoryStore';
+import { VideoProgressMinMax } from '#src/config';
+
+type QueuedProgress = {
+  item: PlaylistItem;
+  progress: number;
+};
+
+/**
+ * The `useWatchHistory` hook has two responsibilities.
+ *
+ * 1. Determine the `startTime` of the given `item` (return value)
+ * 2. Save the watch progress of the current `item` to the user profile
+ *
+ * __The problem:__
+ *
+ * There are multiple events that trigger a save. This results in duplicate (unnecessary) saves and API calls. Another
+ * problem is that some events are triggered when the `item` to update has changed. For example, when clicking a media
+ * item in the "Related shelf". This causes the wrong item to be saved in the watch history.
+ *
+ * __The solution:__
+ *
+ * This hook listens to the player time update event and queues a watch history entry with the current progress and
+ * item. When this needs to be saved, the queue is used to look up the last progress and item and save it when
+ * necessary. The queue is then cleared to prevent duplicate saves and API calls.
+ *
+ * @param player
+ * @param item
+ */
+export const useWatchHistory = (player: JWPlayer | undefined, item: PlaylistItem) => {
+  const queuedWatchProgress = useRef<QueuedProgress | null>(null) as MutableRefObject<QueuedProgress | null>;
+
+  // config
+  const { features } = useConfigStore((s) => s.config);
+  const continueWatchingList = features?.continueWatchingList;
+  const watchHistoryEnabled = !!continueWatchingList;
+
+  // watch History
+  const watchHistoryItem = useWatchHistoryStore((state) => (!!item && watchHistoryEnabled ? state.getItem(item) : undefined));
+
+  // update the queued watch progress on each time update event
+  const handleTimeUpdate = useEventCallback(() => {
+    if (!player) return;
+
+    const progress = player.getPosition() / item.duration;
```

Comment:
Can we get any of this info from the event data? https://docs.jwplayer.com/players/reference/seek-events-1#ontime

It looks like the progress is in the event. Maybe there's a chance that the metadata property will also contain the item being played and then we can have higher confidence that there won't be race conditions.

Reply:
Yes, that's a good observation. I also thought about that, but I couldn't remember why we're using the `item.duration` instead of the player duration.

We can use the event params instead. Initially, I tried using the [getPlaylistItem](https://docs.jwplayer.com/players/reference/getplaylistitem_index_) API to get the current playing item, but JW Player sanitizes the item making it incompatible with the PlaylistItem type we're using in the web app.

I also see an undocumented `absolutePosition` property in the event object. I think this can be used for live events.

Reply:
I just tested a live event:

![image](https://github.com/jwplayer/ott-web-app/assets/3996119/7fbe7aac-f7de-462d-a77f-6b475d8af586)

Reply:
I've updated the event handler